CAT/strategy/NCAT_strategy.py

```diff
@@ -0,0 +1,665 @@
+                student_ids = torch.LongTensor(student_ids).to(device)
+                question_ids = torch.LongTensor(question_ids).to(device)
+                output = self.forward(student_ids, question_ids).view(-1)
+                pred += output.tolist()
+            self.train()
+
+        real = np.array(real)
+        real = np.where(real < 0.5, 0.0, 1.0)
+        pred = np.array(pred)
+        auc = roc_auc_score(real, pred)
+        
+        # Calculate accuracy
+        threshold = 0.5  # You may adjust the threshold based on your use case
+        binary_pred = (pred >= threshold).astype(int)
+        acc = accuracy_score(real, binary_pred)
+
+        return auc,acc
+    
+class env:
+    def __init__(self,data,concept_map,config,T):
+        self.config = config
+        self.T = T
+        self.CDM = 'IRT'
+        self.rates = {}
+        self.users = {}
+        self.utypes = {}
+        #self.args = args
+        self.device = torch.device('cuda')
+        self.rates, self._item_num, self.know_map = self.load_data(data,concept_map)
+        self.tsdata=data
+        self.setup_train_test()
+        self.sup_rates, self.query_rates = self.split_data(ratio=0.5)
+        pth_path='CDM.pt'
+        name = 'IRT'
+        self.model, self.dataset = self.load_CDM(name,data,pth_path,config)
+        #print(self.model)
+    
+    def split_data(self, ratio=0.5):
+        sup_rates, query_rates = {}, {}
+        for u in self.rates:
+            all_items = list(self.rates[u].keys())
+            np.random.shuffle(all_items)
+            sup_rates[u] = {it: self.rates[u][it] for it in all_items[:int(ratio*len(all_items))]}
+            query_rates[u] = {it: self.rates[u][it] for it in all_items[int(ratio*len(all_items)):]}
+        return sup_rates, query_rates
+
+    def re_split_data(self, ratio=0.5):
+        self.sup_rates, self.query_rates = self.split_data(ratio)
+
+    @property
+    def candidate_items(self):
+        return set(self.sup_rates[self.state[0][0]].keys())
+
+    @property
+    def user_num(self):
+        return len(self.rates) + 1
+
+    @property
+    def item_num(self):
+        return self._item_num + 1
+
+    @property
+    def utype_num(self):
+        return len(self.utypes) + 1
+
+    def setup_train_test(self):
+        users = list(range(1, self.user_num))
+        np.random.shuffle(users)
+        self.training, self.validation, self.evaluation = np.split(np.asarray(users), [int(.8 * self.user_num - 1),
+                                                                                       int(.9 * self.user_num - 1)])
+
+    def reset(self):
+        self.reset_with_users(np.random.choice(self.training))
+
+    def reset_with_users(self, uid):
+        self.state = [(uid,1), []]
+        self.short = {}
+        return self.state
+    
+    def load_data(self, ncatdata,concept):
+
+        return ncatdata.data, ncatdata.num_questions, concept
+    def load_CDM(self,name,data,pth_path,config):
```

Comment:
It's better to use some auto tools to check the format of your code. blank line between functions, space after the comma, etc.